compiler/luci/import/src/Nodes/CircleConst.cpp

```diff
@@ -67,6 +67,54 @@ void copy_data(const VectorWrapper<uint8_t> &raw_data, uint32_t num_elements,
   }
 }
 
+template <>
+void copy_data<loco::DataType::S4>(const VectorWrapper<uint8_t> &raw_data, uint32_t num_elements,
+                                   CircleConst *const_node)
+{
+  // TODO support sparse?
+  assert(const_node->sparsityparam() == nullptr);
+  if (const_node->sparsityparam())
+    return;
+
+  uint32_t raw_size = (num_elements + 1) / 2;
+  assert(raw_data.size() == raw_size);
+
+  const uint8_t *data = raw_data.data();
+  const_node->size<loco::DataType::S4>(num_elements);
+  for (uint32_t i = 0; i < raw_size; ++i)
+  {
+    uint32_t idx = i * 2;
+    // 1bit for sign, 3bit for value
+    const_node->at<loco::DataType::S4>(idx++) = static_cast<int8_t>(data[i] << 4) >> 4;
+    if (idx < num_elements)
+      const_node->at<loco::DataType::S4>(idx) = static_cast<int8_t>(data[i]) >> 4;
+  }
+}
+
+template <>
+void copy_data<loco::DataType::U4>(const VectorWrapper<uint8_t> &raw_data, uint32_t num_elements,
+                                   CircleConst *const_node)
+{
+  // TODO support sparse?
+  assert(const_node->sparsityparam() == nullptr);
+  if (const_node->sparsityparam())
+    return;
+
+  uint32_t raw_size = (num_elements + 1) / 2;
+  assert(raw_data.size() == raw_size);
+
+  const uint8_t *data = raw_data.data();
+  const_node->size<loco::DataType::U4>(num_elements);
+  for (uint32_t i = 0; i < raw_size; ++i)
+  {
+    uint32_t idx = i * 2;
+    // 1bit for sign, 3bit for value
+    const_node->at<loco::DataType::U4>(idx++) = static_cast<uint8_t>(data[i] << 4) >> 4;
```

Comment:
Sorry for comment on merged PR.
I'm just curious about `uint8`. If this data is considered as unsigned, I think we do not have to shift twice but just use 
`data[i] & 0xF`

Reply:
#12803 is what I wanted to do someday...